tough/src/key_source.rs

```diff
@@ -12,7 +12,7 @@ use std::result::Result;
 
 /// This trait should be implemented for each source of signing keys. Examples
 /// of sources include: files, AWS SSM, etc.
-pub trait KeySource: Debug + Send + Sync {
+pub trait KeySource: Debug + Send + Sync + KeySourceClone {
```

Comment:
I think this makes the trait backwards incompatible, so we'll need to do an MV bump for this. Is there some way we can avoid adding the additional trait boundary? I had a hard time deciphering why `clap` cares if this is `Clone`-able.

Reply:
I'm guessing this has something to do with `value_parser`...

Reply:
If so, I wonder if we can omit the `value_parser` clarification entirely by making sure that `Box<dyn KeySource>` implements `FromStr`?

```rust
impl std::str::FromStr for Box<dyn KeySource> {
    type Err = crate::error::Error;
    fn from_str(i: &str) -> crate::error::Result<Box<dyn KeySource>> {
        parse_key_source(i)
    }
}
```

Reply:
I'm not sure we can. `parse_key_source` is something in `tuftool` to help figure out what type of KeySource to return. If we move that here, that would require hard coding some of those sub-projects into this one.  Possible, but we would need to figure out a clean way to architect that.

Reply:
The other option could be we just parse this argument into a `String` and do the `parse_key_source` elsewhere in the code. Invalid input wouldn't be handled immediately, but we could do the processing of it early on in the handler so it wouldn't be much different.

Reply:
See #678 for this alternative approach.

Reply:
Sorry for the latency. We could also use a newtype around `KeySource` in tuftool which implemented `FromStr` and then implement `From<ParseableKeySource> for KeySource` in tuftool as well, which I think has all the useful properties of making clap do the validation while not introducing new dependencies.

Reply:
Good idea. I can try that out and see how it goes, but that should work and would make it nice and clean.

Reply:
OK, I've tried a few things, and either this approach won't work, or my Rust skills are too lacking.

The main gotcha I ran into is that the `ParseableKeySource` would also need to implement `Clone`.  If we want to take advantage of the initial Clap argument parsing, that would mean the parsing to `KeySource` would need to happen as part of the `ToString` implementation. Which means the resulting `ParseableKeySource` would need to contain a reference to the parsed `KeySource`.

But then we need to be able to Clone the `ParseableKeySource`, which would imply that `KeySource` also needs to implement the `Clone` trait, which brings us back to the issues here.

The `From` implementation can't return an error if there was a problem parsing the input to get a resulting `Box<dyn KeySource>`, and would get us back to the point where we're just validating the input at the point we actually need the `KeySource`. So unless there's a cleaner way that I'm not seeing (which is quite possible) I think #678 is probably the best option at this point.